atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -0,0 +1,428 @@
+        }
+    }
+
+    public CassandraClientPool(CassandraKeyValueServiceConfig config) {
+        this.config = config;
+        refreshDaemon = PTExecutors.newScheduledThreadPool(1, new ThreadFactoryBuilder().setDaemon(true).setNameFormat("CassandraClientPoolingContainerRefreshDaemon-%d").build());
+        refreshDaemon.scheduleWithFixedDelay(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    refreshPool();
+                } catch (Throwable t) {
+                    log.error("Failed to refresh Cassandra KVS pool. Extended periods of being unable to refresh will cause perf degradation.", t);
+                }
+            }
+        }, CassandraConstants.SECONDS_BETWEEN_GETTING_HOST_LIST, CassandraConstants.SECONDS_BETWEEN_GETTING_HOST_LIST, TimeUnit.SECONDS);
+
+        refreshPool(); // ensure we've initialized before returning
+    }
+
+
+    public void shutdown() {
+        refreshDaemon.shutdown();
+        currentPools.forEach((address, cassandraClientPoolingContainer) -> cassandraClientPoolingContainer.shutdownPooling());
+    }
+
+    private void refreshPool() {
+        checkAndUpdateBlacklist();
+
+        Set<InetSocketAddress> serversToAdd = config.servers();
+        Set<InetSocketAddress> serversToRemove = ImmutableSet.of();
+
+        if (config.autoRefreshNodes()) {
+            refreshTokenRanges(); // re-use token mapping as list of hosts in the cluster
+            for (List<InetSocketAddress> rangeOwners : tokenMap.asMapOfRanges().values()) {
+                for (InetSocketAddress address : rangeOwners) {
+                    serversToAdd.add(address);
+                }
+            }
+        }
+        Sets.difference(Sets.difference(serversToAdd, currentPools.keySet()), blacklistedHosts.keySet());
+
+        if (!config.autoRefreshNodes()) { // (we would just add them back in)
+            serversToRemove = Sets.difference(currentPools.keySet(), config.servers());
+        }
+
+        for (InetSocketAddress newServer : serversToAdd) {
+            currentPools.put(newServer, new CassandraClientPoolingContainer(newServer, config));
+        }
+
+        if (!(serversToAdd.isEmpty() && serversToRemove.isEmpty())) { // if we made any changes
+            sanityCheckRingConsistency();
+            if(!config.autoRefreshNodes()) { // grab new token mapping, if we didn't already do this before
+                refreshTokenRanges();
+            }
+        }
+    }
+
+    private void checkAndUpdateBlacklist() {
+        // Check blacklist and re-integrate or continue to wait as necessary
+        for (Map.Entry<InetSocketAddress, Long> blacklistedEntry : blacklistedHosts.entrySet()) {
+            if (blacklistedEntry.getValue() + BANNED_HOST_BACKOFF_TIME_MS < System.currentTimeMillis()) {
+                InetSocketAddress host = blacklistedEntry.getKey();
+                if(isHostHealthy(host)) {
+                    blacklistedHosts.remove(host);
```

Comment:
This is going to throw concurrent modification exception as you're iterating over the map you're modifying -- probably want to remove via iterator